rpc/gnoi/gnoi_cert.proto

```diff
@@ -0,0 +1,378 @@
+  // succeed. Revoking a certificate should render the existing certificate
+  // unusable by any endpoints.
+  rpc RevokeCertificates(RevokeCertificatesRequest)
+      returns (RevokeCertificatesResponse);
+
+  // An RPC to ask a target if it can generate a Certificate.
+  rpc CanGenerateCsr(CanGenerateCsrRequest) returns (CanGenerateCsrResponse);
+}
+
+// Request messages to rotate existing certificates on the target.
+message RotateCertificateRequest {
+  // Request Messages.
+  oneof rotate_request {
+    GenerateCSRRequest generate_csr = 1;
+    LoadCertificateRequest load_certificate = 2;
+    FinalizeRequest finalize_rotation = 3;
+  }
+}
+
+// Response Messages from the target.
+message RotateCertificateResponse {
+  // Response messages.
+  oneof rotate_response {
+    GenerateCSRResponse generated_csr = 1;
+    LoadCertificateResponse load_certificate = 2;
+  }
+}
+
+// Request messages to install new certificates on the target.
+message InstallCertificateRequest {
+  // Request Messages.
+  oneof install_request {
+    GenerateCSRRequest generate_csr = 1;
+    LoadCertificateRequest load_certificate = 2;
+  }
+}
+
+// Response Messages from the target for the InstallCertificateRequest.
+message InstallCertificateResponse {
+  // Response messages.
+  oneof install_response {
+    GenerateCSRResponse generated_csr = 1;
+    LoadCertificateResponse load_certificate = 2;
+  }
+}
+
+// Request to generate the CSR.
+// When this request is made for rotating an existing certificate as part of the
+// Rotate() RPC, then the target must ensure that the "certificate_id" is
+// already created and exists on the target. If the Certificate Rotation
+// proceeds to load the certificate, it must associate the new certificate with
+// the previously created "certificate_id".
+//
+// When this request is made for installing a completely new certificate as part
+// of the Install() RPC , then the target must ensure that the "certificate_id"
+// is completely new and no entities on the target are should be bound to this
+// certificate_id. If any existing certificate matches the certificate_id, then
+// this request should fail.
+//
+// If there is another ongoing Rotate/Install RPC with the same certificate_id,
+// the GenerateCSRRequest should fail.
+// should fail.
+message GenerateCSRRequest {
+  // Parameters for creating a CSR.
+  CSRParams csr_params = 1;
+  // The certificate id with which this CSR will be associated. The target
+  // configuration should bind an entity which wants to use a certificate to
+  // the certificate_id it should use.
+  string certificate_id = 2;
+}
+
+// Parameters to be used when generating a Certificate Signing Request.
+message CSRParams {
+  // The type of certificate which will be will be associated for this CSR.
+  CertificateType type = 1;
+
+  // Minimum size of the key to be used by the target when generating a
+  // public/private key pair.
+  uint32 min_key_size = 2;
```

Comment:
Should this rather be the actual key size that one wants to request? It seems strange to just give a minimum - for example, if I specify 1024 here, should the target generate a 2048 or 16384 sized key?